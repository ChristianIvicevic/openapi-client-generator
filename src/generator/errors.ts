/* istanbul ignore file */

// eslint-disable-next-line max-classes-per-file
import type { OpenAPIV3 } from 'openapi-types';

export class UnsupportedMixedSchemaObjectError extends Error {
  public constructor(private readonly schemaName: string) {
    super(
      `Unsupported mixed schema object error for '${schemaName}'. The type of '${schemaName}' is declared as an array, however the generator does not support this. Did you mean to make the type nullable according to OpenAPI v3.1? If so, use 'oneOf' including 'type: null' as one of the possible types instead.`,
    );
  }
}

export class ReferenceDoesntMatchValidPatternError extends Error {
  public constructor() {
    super('Cannot dereference specified ref.');
  }
}

export class OperationIdNotDefinedError extends Error {
  public constructor(
    private readonly path: string,
    private readonly method: OpenAPIV3.HttpMethods,
  ) {
    super(
      `The '${method}' method for the path '${path}' needs an operation id.`,
    );
  }
}

export class OperationHasNoResponsesError extends Error {
  public constructor(
    private readonly path: string,
    private readonly method: OpenAPIV3.HttpMethods,
  ) {
    super(`The '${method}' method for the path '${path}' has no responses.`);
  }
}
