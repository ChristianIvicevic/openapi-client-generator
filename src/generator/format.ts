import { pipe } from 'fp-ts/lib/function';
import * as RA from 'fp-ts/ReadonlyArray';
import { format as prettier } from 'prettier';
import ts from 'typescript';

/**
 * Formats the specified code using prettier.
 * @param code Code to format.
 */
export const formatUsingPrettier = (code: string) =>
  prettier(code, {
    ...{
      arrowParens: 'avoid',
      singleQuote: true,
      tabWidth: 2,
      trailingComma: 'all',
    },
    parser: 'typescript',
  });

/**
 * Organizes and removes unused imports from the specified code.
 * @param code Code to organize.
 */
export const organizeImports = (code: string) => {
  const languageServiceHost = new InMemoryLanguageServiceHost();
  const languageService = ts.createLanguageService(
    languageServiceHost,
    ts.createDocumentRegistry(),
  );
  languageServiceHost.addFile('output.ts', code);
  const fileTextChanges = languageService.organizeImports(
    {
      type: 'file',
      fileName: 'output.ts',
    },
    ts.getDefaultFormatCodeSettings(),
    undefined,
  );

  return fileTextChanges.length === 0
    ? code
    : applyTextChanges(code, fileTextChanges[0].textChanges);
};

const applyTextChanges = (content: string, changes: readonly ts.TextChange[]) =>
  pipe(
    changes,
    RA.reverse,
    RA.reduce(content, (newContent, { span, newText }) =>
      [
        newContent.slice(0, span.start),
        newText,
        newContent.slice(span.start + span.length),
      ].join(''),
    ),
  );

/* istanbul ignore next */
class InMemoryLanguageServiceHost implements ts.LanguageServiceHost {
  private files: Record<string, { file: ts.IScriptSnapshot; ver: number }> = {};

  public getCompilationSettings = ts.getDefaultCompilerOptions;

  public getScriptFileNames(): string[] {
    return Object.keys(this.files);
  }

  public getScriptVersion(fileName: string): string {
    return String(this.files[fileName].ver);
  }

  public getScriptSnapshot(fileName: string): ts.IScriptSnapshot | undefined {
    return this.files[fileName]?.file;
  }

  // eslint-disable-next-line class-methods-use-this
  public getCurrentDirectory(): string {
    return '';
  }

  // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars
  public getDefaultLibFileName(options: ts.CompilerOptions): string {
    return 'lib';
  }

  public addFile(fileName: string, content: string) {
    const snapshot = ts.ScriptSnapshot.fromString(content);
    snapshot.getChangeRange = () => undefined;
    if (this.files[fileName] === undefined) {
      this.files[fileName] = { ver: 1, file: snapshot };
    } else {
      this.files[fileName].ver += 1;
      this.files[fileName].file = snapshot;
    }
  }
}
