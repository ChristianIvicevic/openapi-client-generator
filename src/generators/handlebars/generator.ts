import { pascalCase } from 'change-case';
import { generateSchema } from 'generators/handlebars/schemas';
import type {
  RequestsCompilationContext,
  SchemasCompilationContext,
} from 'generators/handlebars/types';
import * as Handlebars from 'handlebars';
import type {
  DocumentInfo,
  OperationInfo,
  OperationMethod,
  ParameterInfo,
  TypeInfo,
} from 'parser/types';
import { format } from 'prettier';
import { getLogger } from 'utils/logging';
import wordWrap from 'word-wrap';

const wrapJsdoc = (text: string) =>
  wordWrap(text, {
    width: 80 - 3,
    newline: '\n * ',
    indent: '',
  });

// BEGIN Handlebars helper functions

const formatParameters = (text: string) => text.replace(/\{/g, '${parameters.');

const getRequestBodyArgument = (
  method: OperationMethod,
  typeInfo?: readonly TypeInfo[],
) => {
  if (method !== 'post' && method !== 'put') {
    return '';
  }

  return typeInfo === undefined ? 'undefined,' : 'requestBody,';
};

const hasQueryParameters = (parameterInfo: ParameterInfo[]) =>
  parameterInfo.some(parameter => parameter.in === 'query');

const hasRequestBody = (typeInfo?: readonly TypeInfo[]) =>
  typeInfo !== undefined;

const isQueryParameter = (parameterInfo: ParameterInfo) =>
  parameterInfo.in === 'query';

const resolveType = ({ isArray, type }: TypeInfo) =>
  isArray ? `readonly ${type}[]` : type;

const resolveTypes = (typeInfo?: readonly TypeInfo[]) => {
  const logger = getLogger();

  // TODO: Figure out whether this is even possible.
  if (typeInfo === undefined) {
    logger.warn(`A type was unexpectedly resolved to 'unknown'`);
    return 'unknown';
  }

  const compiledType = typeInfo.map(resolveType).join(' | ');
  return `(${compiledType})`;
};

Handlebars.registerHelper('format-parameters', formatParameters);
Handlebars.registerHelper('get-request-body-arg', getRequestBodyArgument);
Handlebars.registerHelper('generate-schema', generateSchema);
Handlebars.registerHelper('has-query-parameters', hasQueryParameters);
Handlebars.registerHelper('has-request-body', hasRequestBody);
Handlebars.registerHelper('is-query-parameter', isQueryParameter);
Handlebars.registerHelper('resolve-type', resolveType);
Handlebars.registerHelper('resolve-types', resolveTypes);
Handlebars.registerHelper('wrap-jsdoc', wrapJsdoc);

// END Handlebars helper functions

const prettier = (content: string) =>
  format(content, {
    ...{
      arrowParens: 'avoid',
      singleQuote: true,
      tabWidth: 2,
      trailingComma: 'all',
    },
    parser: 'typescript',
  });

export const compileUsingHandlebars = async ({
  pathItemObjects,
  schemaObjects,
}: DocumentInfo) => {
  // The import MUST have the explicit suffix otherwise the ts path plugin
  // won't be able to resolve it properly for runtime
  await import('generators/handlebars/templates/templates.precompiled.js');

  const schemasTemplate = Handlebars.templates[
    'schemas.hbs'
  ] as HandlebarsTemplateDelegate<SchemasCompilationContext>;

  const requestsTemplate = Handlebars.templates[
    'requests.hbs'
  ] as HandlebarsTemplateDelegate<RequestsCompilationContext>;

  const operations = Object.values(pathItemObjects)
    .map(
      pathItemObject =>
        Object.values(pathItemObject).filter(Boolean) as OperationInfo[],
    )
    .flat()
    .sort((a, b) => a.operationId.localeCompare(b.operationId));

  const referencedTypes = operations.flatMap(
    ({ parameters, requestBodyType, responseType }) => [
      ...parameters.map(({ typeInfo: { type } }) => type),
      ...(requestBodyType?.map(({ type }) => type) ?? []),
      ...responseType.map(({ type }) => type),
    ],
  );

  const referencedSchemas = [
    ...new Set(
      referencedTypes.filter(
        type =>
          ![
            'boolean',
            'number',
            'object',
            'string',
            'unknown',
            'void',
          ].includes(type),
      ),
    ),
  ].sort((a, b) => a.localeCompare(b));

  const parsedSchemaObjects = Object.entries(schemaObjects)
    .map(([name, schemaObject]) => ({ name: pascalCase(name), schemaObject }))
    .sort((a, b) => a.name.localeCompare(b.name));

  const compiledSchemasOutput = schemasTemplate({
    schemaObjects: parsedSchemaObjects,
  });

  const compiledRequestsOutput = requestsTemplate({
    operations,
    referencedSchemas,
  });

  return {
    schemas: prettier(compiledSchemasOutput),
    requests: prettier(compiledRequestsOutput),
  };
};
