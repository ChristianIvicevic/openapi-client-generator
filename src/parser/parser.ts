import { camelCase } from 'change-case';
import type { OpenAPIV3 } from 'openapi-types';
import { resolveContentType, resolveParameter } from 'parser/resolver';
import type {
  DocumentInfo,
  OperationInfo,
  OperationMethod,
  ParserContext,
  PathItemObjectInfo,
} from 'parser/types';
import { partial } from 'ramda';
import { assertIsDefined } from 'utils/assert';
import { getLogger } from 'utils/logging';
import yaml from 'yaml';

export const parseYaml = (content: string): DocumentInfo => {
  const logger = getLogger();
  const document = yaml.parse(content) as OpenAPIV3.Document;

  const schemaObjects = Object.entries(
    document.components?.schemas ?? {},
  ).reduce<
    Record<
      string,
      | OpenAPIV3.ReferenceObject
      | OpenAPIV3.ArraySchemaObject
      | OpenAPIV3.NonArraySchemaObject
    >
  >(
    (currentSchemas, [schemaName, schemaObject]) => ({
      ...currentSchemas,
      [schemaName]: schemaObject,
    }),
    {},
  );

  const pathItemObjects = Object.entries(document.paths ?? {}).map(
    ([path, pathItemObject]) => {
      assertIsDefined(pathItemObject);

      logger.verbose(`Parsing path '${path}'`);

      const context: ParserContext = { document, path };
      return parsePathItemObject(context, pathItemObject);
    },
  );

  return { schemaObjects, pathItemObjects };
};

const parsePathItemObject = (
  context: ParserContext,
  pathItemObject: OpenAPIV3.PathItemObject,
): PathItemObjectInfo => {
  const parameters = (pathItemObject.parameters ?? []).map(
    partial(resolveParameter, [context]),
  );

  // TODO: Use TS plugin to generate from type
  return ([
    'get',
    'post',
    'put',
    'delete',
  ] as OperationMethod[]).reduce<PathItemObjectInfo>(
    (acc, method) => ({
      ...acc,
      [method]: parseOperationObject(
        { ...context, method, parameters },
        pathItemObject[method],
      ),
    }),
    {},
  );
};

const parseOperationObject = (
  context: ParserContext,
  operationObject?: OpenAPIV3.OperationObject,
): OperationInfo | undefined => {
  const logger = getLogger();
  const { path, method, parameters } = context;

  assertIsDefined(path);
  assertIsDefined(method);

  if (operationObject === undefined) {
    logger.debug(`Skipping empty '${method}' operation object`);
    return undefined;
  }

  if (operationObject.operationId === undefined) {
    throw Error(`'${method}' operation object has no operation id`);
  }

  logger.debug(`Parsing responses for '${method}' operation object`);

  const { responses, requestBody, operationId, summary } = operationObject;

  const responseType = resolveContentType(
    Object.entries(responses ?? {})
      .filter(isResponseOk)
      .map(([, response]) => response),
  );

  // TODO: Handle required flag for the request body
  const requestBodyType =
    requestBody === undefined ? undefined : resolveContentType([requestBody]);

  const operationParameters = [
    ...(parameters ?? []),
    ...(operationObject.parameters ?? []).map(
      partial(resolveParameter, [context]),
    ),
  ];

  return {
    path,
    method,
    operationId: camelCase(operationId),
    summary,
    responseType,
    requestBodyType,
    parameters: operationParameters,
  };
};

const isResponseOk = ([statusCode]: [string, unknown]) =>
  statusCode.toString().startsWith('2') ||
  // We have to consider 'default' responses as well since they can be of any
  // kind, be it OK or ERROR
  statusCode.toString().startsWith('default');
