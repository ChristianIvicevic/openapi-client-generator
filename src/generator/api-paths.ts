import * as E from 'fp-ts/Either';
import { flow, pipe } from 'fp-ts/function';
import * as O from 'fp-ts/Option';
import * as RA from 'fp-ts/ReadonlyArray';
import * as R from 'fp-ts/Record';
import { Constants } from 'generator/constants';
import { OperationIdNotDefinedError } from 'generator/errors';
import { formatUsingPrettier, organizeImports } from 'generator/format';
import {
  createPathTemplateExpression,
  generateParameterSignatures,
} from 'generator/operations';
import type { GeneratorOptions } from 'generator/options';
import {
  applicativeValidation,
  assertIsDefined,
  compileNodes,
  liftN,
} from 'generator/utils';
import type { OpenAPIV3_1 } from 'openapi-types';
import { OpenAPIV3 } from 'openapi-types';
import ts, { factory } from 'typescript';

const supportedMethods = Object.values(OpenAPIV3.HttpMethods);

// TODO: Extract this block of code into a generic path traversal function.
export const generateApiPaths = (
  document: OpenAPIV3_1.Document,
  options?: GeneratorOptions,
) =>
  pipe(
    document.paths ?? {},
    R.toArray,
    RA.map(([path, pathItemObject]) => {
      assertIsDefined(pathItemObject);
      return generateApiPathNodes(document, pathItemObject, path);
    }),
    RA.sequence(applicativeValidation),
    E.mapLeft(RA.map(error => error.message)),
    E.map(RA.flatten),
    E.map(propertyAssignments => [
      factory.createVariableStatement(
        [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        factory.createVariableDeclarationList(
          [
            factory.createVariableDeclaration(
              factory.createIdentifier(Constants.API_PATHS_OBJECT_NAME),
              undefined,
              undefined,
              factory.createObjectLiteralExpression(
                propertyAssignments as ts.ObjectLiteralElementLike[],
                true,
              ),
            ),
          ],
          ts.NodeFlags.Const,
        ),
      ),
    ]),
    E.map(flow(compileNodes, organizeImports, formatUsingPrettier)),
  );

const generateApiPathNodes = (
  document: OpenAPIV3_1.Document,
  pathItemObject: OpenAPIV3_1.PathItemObject,
  path: string,
) =>
  pipe(
    Object.entries(pathItemObject),
    RA.filterMap(([method, operationObject]) =>
      supportedMethods.includes(method as OpenAPIV3.HttpMethods)
        ? O.some([
            method as OpenAPIV3.HttpMethods,
            // This type narrowing seems to be required since the inferred type
            // for the operation object is bigger than expected.
            operationObject as OpenAPIV3_1.OperationObject,
          ] as const)
        : O.none,
    ),
    RA.map(([method, operationObject]) =>
      liftN(generateOperationObjectNode)(
        document,
        path,
        pathItemObject.parameters,
        method,
        operationObject,
      ),
    ),
    RA.sequence(applicativeValidation),
  );

const generateOperationObjectNode = (
  document: OpenAPIV3_1.Document,
  path: string,
  parameters: OpenAPIV3_1.PathItemObject['parameters'],
  method: OpenAPIV3_1.HttpMethods,
  operationObject: OpenAPIV3_1.OperationObject,
): E.Either<Error, ts.Node> => {
  if (operationObject.operationId === undefined) {
    return E.left(new OperationIdNotDefinedError(path, method));
  }

  const parametersParameterDeclaration = pipe(
    generateParameterSignatures(
      document,
      'path',
      parameters,
      operationObject.parameters,
    ),
    O.map(propertySignature =>
      factory.createParameterDeclaration(
        undefined,
        undefined,
        undefined,
        factory.createIdentifier(Constants.PARAMETERS_PARAMETER_NAME),
        undefined,
        factory.createTypeLiteralNode(propertySignature),
        undefined,
      ),
    ),
  );

  const apiPath = (
    O.isSome(parametersParameterDeclaration)
      ? createPathTemplateExpression
      : factory.createStringLiteral
  )(path);

  const propertyAssignment = factory.createPropertyAssignment(
    factory.createIdentifier(operationObject.operationId),
    factory.createArrowFunction(
      undefined,
      undefined,
      RA.compact([parametersParameterDeclaration]),
      undefined,
      factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
      apiPath,
    ),
  );

  if (operationObject.summary !== undefined) {
    ts.addSyntheticLeadingComment(
      propertyAssignment,
      ts.SyntaxKind.MultiLineCommentTrivia,
      `*\n * ${operationObject.summary}\n `,
      true,
    );
  }

  return E.right(propertyAssignment);
};
