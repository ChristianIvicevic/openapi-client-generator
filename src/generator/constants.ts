import { OperationMethod } from 'generator/types';

export const Constants = {
  AXIOS_IMPORT_NAME: 'axios',
  AXIOS_PACKAGE_NAME: 'axios',
  AXIOS_REQUEST_CONFIG_PARAMETER_NAME: 'config',
  AXIOS_REQUEST_CONFIG_TYPE: 'AxiosRequestConfig',
  PARAMETERS_PARAMETER_NAME: 'parameters',
  REQUEST_BODY_PARAMETER_NAME: 'requestBody',
  SUPPORTED_MEDIA_TYPES: [
    '*/*',
    'application/json',
    'application/octet-stream',
    'application/x-yaml',
  ],
  SUPPORTED_OPERATION_METHODS: [
    'get',
    'post',
    'put',
    'delete',
  ] as readonly OperationMethod[],
};
