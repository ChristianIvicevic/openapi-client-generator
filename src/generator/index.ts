import { Constants } from 'generator/constants';
import { createLeadingTrivia } from 'generator/factories/leading-trivia';
import type { CreateOperationOrThrowParameters } from 'generator/factories/operation';
import { createOperationOrThrow } from 'generator/factories/operation';
import { createSchemaDeclaration } from 'generator/factories/schema-declaration';
import { CompilerOptions } from 'generator/types';
import type { OpenAPIV3 } from 'openapi-types';
import ts, { factory } from 'typescript';
import { assertIsDefined } from 'utils/assert';
import { format, organizeImports } from 'utils/format';
import { compact } from 'utils/fp';
import winston from 'winston';

export const compileDocument = (
  document: OpenAPIV3.Document,
  options?: CompilerOptions,
) => {
  // TODO: Evaluate whether we can drop tracking of referenced schemas.
  const referencedSchemas: string[] = [];

  winston.verbose('Compiling schemas...');
  const compiledSchemas = Object.entries(document.components?.schemas ?? {})
    .sort(([a], [b]) => a.localeCompare(b))
    .flatMap(([schema, schemaObject]) => {
      winston.debug(`Compiling schema '${schema}'`);
      return createSchemaDeclaration(
        { document, referencedSchemas },
        schema,
        schemaObject,
      );
    });

  winston.verbose('Compiling operations...');
  // This block is a workaround to prefetch operation ids in order to sort the
  // generated methods alphanumerically rather than in their order inside the
  // schema file. This is intentional to produce more deterministic output
  // during compilation.
  const compiledOperations = compact(
    Object.entries(document.paths ?? {})
      .flatMap(([path, pathItemObject]) => {
        assertIsDefined(pathItemObject);

        const { parameters: pathParameters } = pathItemObject;

        return compact(
          Constants.SUPPORTED_OPERATION_METHODS.map((method):
            | CreateOperationOrThrowParameters
            | undefined => {
            const operationObject = pathItemObject[method];

            if (operationObject === undefined) {
              winston.debug(
                `Skipping empty '${method}' operation on path '${path}'`,
              );
              return undefined;
            }

            return {
              operationId: operationObject.operationId ?? '<NO OPERATION ID>',
              parameters: [
                { document, path, referencedSchemas },
                method,
                pathParameters,
                operationObject,
              ],
            };
          }),
        );
      })
      .sort((a, b) => a.operationId.localeCompare(b.operationId))
      .flatMap(({ parameters }) => createOperationOrThrow(...parameters)),
  );

  const schemasContent = compileNodes(compiledSchemas);
  const requestsContent = compileNodes([
    ...createLeadingTrivia({
      referencedSchemas: [
        ...new Set(referencedSchemas.sort((a, b) => a.localeCompare(b))),
      ],
      schemasFileName: options?.schemasFileName,
    }),
    ...compiledOperations,
  ]);

  return {
    schemas: format(schemasContent),
    requests: format(organizeImports(requestsContent)),
  };
};

const compileNodes = <T extends ts.Node>(elements: readonly T[]) =>
  [
    '/* eslint-disable */',
    '/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */',
    ts
      .createPrinter({ newLine: ts.NewLineKind.LineFeed })
      .printList(
        ts.ListFormat.MultiLine,
        factory.createNodeArray(elements),
        ts.createSourceFile(
          'output.ts',
          '',
          ts.ScriptTarget.Latest,
          false,
          ts.ScriptKind.TS,
        ),
      ),
  ].join('\n');
