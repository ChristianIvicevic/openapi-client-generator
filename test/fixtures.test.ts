import { readFileSync } from 'fs';
import { compileDocument } from 'generator';
import { parseYaml } from 'parser';
import { compose } from 'ramda';

// TESTEE function
const compile = compose(compileDocument, parseYaml);

describe('Fixtures', () => {
  it('compiles the link example schema', () => {
    // GIVEN the link example OpenAPI schema
    const document = readFileSync('test/fixtures/link-example.yaml').toString();

    // WHEN compiling
    const { requests, schemas } = compile(document);

    // THEN the output matches the snapshot
    expect(requests).toMatchSnapshot();
    expect(schemas).toMatchSnapshot();
  });

  it('compiles the petstore schema', () => {
    // GIVEN the petstore OpenAPI schema
    const document = readFileSync('test/fixtures/petstore.yaml').toString();

    // WHEN compiling
    const { requests, schemas } = compile(document);

    // THEN the output matches the snapshot
    expect(requests).toMatchSnapshot();
    expect(schemas).toMatchSnapshot();
  });

  it('compiles the expanded petstore schema', () => {
    // GIVEN the expanded petstore OpenAPI schema
    const document = readFileSync(
      'test/fixtures/petstore-expanded.yaml',
    ).toString();

    // WHEN compiling
    const { requests, schemas } = compile(document);

    // THEN the output matches the snapshot
    expect(requests).toMatchSnapshot();
    expect(schemas).toMatchSnapshot();
  });

  it('compiles the uspto schema', () => {
    // GIVEN the uspto OpenAPI schema
    const document = readFileSync('test/fixtures/uspto.yaml').toString();

    // WHEN compiling
    const { requests, schemas } = compile(document);

    // THEN the output matches the snapshot
    expect(requests).toMatchSnapshot();
    expect(schemas).toMatchSnapshot();
  });
});
