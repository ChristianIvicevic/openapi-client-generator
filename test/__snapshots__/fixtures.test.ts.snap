// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fixtures compiles the expanded petstore schema 1`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */
import type { AxiosRequestConfig } from 'axios';
import axios from 'axios';
import type { Error, NewPet, Pet } from './schemas';
/**
 * @param requestBody The HTTP request content sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const addPet = async (
  requestBody: NewPet,
  config?: AxiosRequestConfig,
) => axios.post<Pet | Error>('/pets', requestBody, config);
/**
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const deletePet = async (
  parameters: {
    /**
     * ID of pet to delete
     */
    readonly id: number;
  },
  config?: AxiosRequestConfig,
) => axios.delete<void | Error>(\`/pets/\${parameters.id}\`, config);
/**
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const findPetById = async (
  parameters: {
    /**
     * ID of pet to fetch
     */
    readonly id: number;
  },
  config?: AxiosRequestConfig,
) => axios.get<Pet | Error>(\`/pets/\${parameters.id}\`, config);
/**
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const findPets = async (
  parameters: {
    /**
     * tags to filter by
     */
    readonly tags?: readonly string[];
    /**
     * maximum number of results to return
     */
    readonly limit?: number;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<readonly Pet[] | Error>('/pets', {
    ...config,
    params: {
      tags: parameters.tags,
      limit: parameters.limit,
    },
  });
"
`;

exports[`Fixtures compiles the expanded petstore schema 2`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */
export type Error = {
  readonly code: number;
  readonly message: string;
};
export type NewPet = {
  readonly name: string;
  readonly tag?: string;
};
export type Pet = NewPet & {
  readonly id: number;
};
"
`;

exports[`Fixtures compiles the link example schema 1`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */
import type { AxiosRequestConfig } from 'axios';
import axios from 'axios';
import type { Pullrequest, Repository, User } from './schemas';
/**
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const getPullRequestsById = async (
  parameters: {
    readonly username: string;
    readonly slug: string;
    readonly pid: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<Pullrequest>(
    \`/2.0/repositories/\${parameters.username}/\${parameters.slug}/pullrequests/\${parameters.pid}\`,
    config,
  );
/**
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const getPullRequestsByRepository = async (
  parameters: {
    readonly username: string;
    readonly slug: string;
    readonly state?: 'open' | 'merged' | 'declined';
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<readonly Pullrequest[]>(
    \`/2.0/repositories/\${parameters.username}/\${parameters.slug}/pullrequests\`,
    {
      ...config,
      params: {
        state: parameters.state,
      },
    },
  );
/**
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const getRepositoriesByOwner = async (
  parameters: {
    readonly username: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<readonly Repository[]>(
    \`/2.0/repositories/\${parameters.username}\`,
    config,
  );
/**
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const getRepository = async (
  parameters: {
    readonly username: string;
    readonly slug: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<Repository>(
    \`/2.0/repositories/\${parameters.username}/\${parameters.slug}\`,
    config,
  );
/**
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const getUserByName = async (
  parameters: {
    readonly username: string;
  },
  config?: AxiosRequestConfig,
) => axios.get<User>(\`/2.0/users/\${parameters.username}\`, config);
/**
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const mergePullRequest = async (
  parameters: {
    readonly username: string;
    readonly slug: string;
    readonly pid: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.post<void>(
    \`/2.0/repositories/\${parameters.username}/\${parameters.slug}/pullrequests/\${parameters.pid}/merge\`,
    undefined,
    config,
  );
"
`;

exports[`Fixtures compiles the link example schema 2`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */
export type Pullrequest = {
  readonly id?: number;
  readonly title?: string;
  readonly repository?: Repository;
  readonly author?: User;
};
export type Repository = {
  readonly slug?: string;
  readonly owner?: User;
};
export type User = {
  readonly username?: string;
  readonly uuid?: string;
};
"
`;

exports[`Fixtures compiles the petstore schema 1`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */
import type { AxiosRequestConfig } from 'axios';
import axios from 'axios';
import type { Error, Pet, Pets } from './schemas';
/**
 * Create a pet
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const createPets = async (config?: AxiosRequestConfig) =>
  axios.post<void | Error>('/pets', undefined, config);
/**
 * List all pets
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const listPets = async (
  parameters: {
    /**
     * How many items to return at one time (max 100)
     */
    readonly limit?: number;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<Pets | Error>('/pets', {
    ...config,
    params: {
      limit: parameters.limit,
    },
  });
/**
 * Info for a specific pet
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const showPetById = async (
  parameters: {
    /**
     * The id of the pet to retrieve
     */
    readonly petId: string;
  },
  config?: AxiosRequestConfig,
) => axios.get<Pet | Error>(\`/pets/\${parameters.petId}\`, config);
"
`;

exports[`Fixtures compiles the petstore schema 2`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */
export type Error = {
  readonly code: number;
  readonly message: string;
};
export type Pet = {
  readonly id: number;
  readonly name: string;
  readonly tag?: string;
};
export type Pets = readonly Pet[];
"
`;

exports[`Fixtures compiles the uspto schema 1`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */
import type { AxiosRequestConfig } from 'axios';
import axios from 'axios';
import type { DataSetList } from './schemas';
/**
 * List available data sets
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const listDataSets = async (config?: AxiosRequestConfig) =>
  axios.get<DataSetList>('/', config);
/**
 * Provides the general information about the API and the list of fields that can be used to query the dataset.
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const listSearchableFields = async (
  parameters: {
    /**
     * Name of the dataset.
     */
    readonly dataset: string;
    /**
     * Version of the dataset.
     */
    readonly version: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<string>(
    \`/\${parameters.dataset}/\${parameters.version}/fields\`,
    config,
  );
/**
 * Provides search capability for the data set with the given search criteria.
 * @param parameters The HTTP request (path, query, header and cookie) parameters sent to the server.
 * @param requestBody The HTTP request content sent to the server.
 * @param config A custom \`AxiosRequestConfig\` object that is used to override the global configuration for this request. This value is optional.
 */
export const performSearch = async (
  parameters: {
    /**
     * Version of the dataset.
     */
    readonly version: string;
    /**
     * Name of the dataset. In this case, the default value is oa_citations
     */
    readonly dataset: string;
  },
  requestBody?: unknown | undefined,
  config?: AxiosRequestConfig,
) =>
  axios.post<readonly Record<string, unknown>[]>(
    \`/\${parameters.dataset}/\${parameters.version}/records\`,
    requestBody,
    config,
  );
"
`;

exports[`Fixtures compiles the uspto schema 2`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */
export type DataSetList = {
  readonly total?: number;
  readonly apis?: readonly {
    readonly apiKey?: string;
    readonly apiVersionNumber?: string;
    readonly apiUrl?: string;
    readonly apiDocumentationUrl?: string;
  }[];
};
"
`;
